src/Processors/Transforms/PartialSortingTransform.cpp

```diff
@@ -19,7 +52,42 @@ void PartialSortingTransform::transform(Chunk & chunk)
     auto block = getInputPort().getHeader().cloneWithColumns(chunk.detachColumns());
     chunk.clear();
 
+    ColumnRawPtrs block_columns;
+    UInt64 rows_num = block.rows();
+
+    if (!threshold_block_columns.empty())
+    {
+        IColumn::Filter filter(rows_num, 0);
+        block_columns = extractColumns(block, description);
+        size_t filtered_count = 0;
+
+        for (UInt64 i = 0; i < rows_num; ++i)
+        {
+            if (less(threshold_block_columns, limit - 1, block_columns, i, description))
+            {
+                ++filtered_count;
+                filter[i] = 1;
+            }
+        }
+
+        if (filtered_count)
+        {
+            for (auto & column : block.getColumns())
+            {
+                column = column->filter(filter, filtered_count);
+            }
+        }
+    }
+
     sortBlock(block, description, limit);
+
+    if (limit && limit < rows_num &&
+        (threshold_block_columns.empty() || less(block_columns, limit - 1, threshold_block_columns, limit - 1, description)))
```

Comment:
We don't call `extractColumns()` if `threshold_block_columns` is empty. That's why tests failed.